From 9869614ead6245808ca8f0763a07a00ebe3ac6b8 Mon Sep 17 00:00:00 2001
From: Ronan Abhamon <ronan.abhamon@vates.fr>
Date: Tue, 4 Oct 2022 18:48:09 +0200
Subject: [PATCH 082/178] fix(LinstorSR): start correctly HA servers (HTTP/NBD)
 after reboot

Use a timeout call after a reboot to get a XAPI session because
it can be impossible to initialize the connection at startup
or if the current host has been ejected (in this case the call can
block indefinitely).

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>
---
 drivers/LinstorSR.py | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/drivers/LinstorSR.py b/drivers/LinstorSR.py
index d32771f..ae25385 100755
--- a/drivers/LinstorSR.py
+++ b/drivers/LinstorSR.py
@@ -2513,7 +2513,10 @@ class LinstorVDI(VDI.VDI):
                 port = '8077'
 
             try:
-                session = util.get_localAPI_session()
+                # Use a timeout call because XAPI may be unusable on startup
+                # or if the host has been ejected. So in this case the call can
+                # block indefinitely.
+                session = util.timeout_call(5, util.get_localAPI_session)
                 host_ip = util.get_this_host_address(session)
             except:
                 # Fallback using the XHA file if session not available.
@@ -2583,7 +2586,7 @@ class LinstorVDI(VDI.VDI):
                 port = '8077'
 
             try:
-                session = util.get_localAPI_session()
+                session = util.timeout_call(5, util.get_localAPI_session)
                 ips = util.get_host_addresses(session)
             except Exception as e:
                 _, ips = get_ips_from_xha_config_file()
-- 
2.45.2

